### PR TITLE
[bug-fix] Call transfer_batch_to_device in DDPlugin  

### LINKED ISSUE #2350: `transfer_batch_to_device` doesn't work under DP
## 🐛 Bug

This is discussed under #1756 and I'm opening a separate issue here for visibility.

In the training loop, for DP/DDP/DDP2, we do not move the data to devices ourselves, but instead use the default scatter to transfer data. This results in `transfer_batch_to_device` not being called.

https://github.com/PyTorchLightning/pytorch-lightning/blob/16a7326e5259a3cdd20a508c34a0f84806d88f8e/pytorch_lightning/trainer/training_loop.py#L736-L737

### Expected behavior

Ideally, we want `transfer_batch_to_device` to work in all settings. If it's not possible at all to override this behavior, at least a run-time warning and/or some warning in the doc should be given.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR moves transfer_batch_to_device from `ddp_sharded` to `ddp_plugin` in `on_before_forward`

Partially solves #2350

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/ddp2_accelerator.py [hunk 1]
```diff
@@ -113,6 +113,8 @@ def init_device(self, process_idx):
 
     def model_to_device(self, model):
         model.cuda(self.trainer.root_gpu)
+        for param in model.parameters():
+            param = param.cuda(self.trainer.root_gpu)
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/ddp2_accelerator.py:117]: why? this is not correct, it will not move buffers anyway :)

**Comment by SeanNaren** [comment]
So a couple of things:

`on_before_forward` is moving towards hook based functionality across the plugins. This I think is preferred and in the original PR we moved from `move_to_device` to `on_before_forward`. It allows the plugins to decide what logic is appropriate before the forward function.

We should not enforce move to device. This wouldn't work in cases where custom distributed accelerators need to move device differently imo. It should be up to the plugin to do what's necessary. Also when using DDP we do not need to manually move tensors to the appropriate device?

**Comment by SeanNaren** [comment]
@tchaton can you have a look at the above comment? I don't think we should merge this.

**Comment by awaelchli** [comment]
Since in Lightning we only ever use the DDP with a single device (in each subprocess), then calling this hook will work (I can help test). However it is not applicable for the general way pytorch DDP can work and it can not be done the same way for the other modes (DP/DDP2).

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 2]
```diff
@@ -91,23 +92,23 @@ def init_ddp_connection(
...
         Override to handle custom input to device logic. For DDP, no logic is required as this is handled internally
         within the DDP wrapper.
 
         Example::
 
-            def on_before_forward(self, model, *args):
-                batch, batch_idx = args
+            def on_before_forward(self, model, *batch):
+                batch, batch_idx = batch
                 return batch.to(model.device)
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/ddp_plugin.py:104]: input is `batch, batch_idx`, but it returns only `batch` in this example?

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 3]
```diff
@@ -91,23 +92,23 @@ def init_ddp_connection(
...
 
         Args:
-            args: Inputs to the model.
+            batch: Inputs to the model.
             model: Model to train.
-        Returns: args moved to correct device if needed.
+        Returns: batch moved to correct device if needed.
         """
-        return args
+        return model.transfer_batch_to_device(batch, model.device)
```
Comment (Borda) [inline pytorch_lightning/plugins/ddp_plugin.py:111]: not very sure about passing this arg as it came from * so it can contain a few args, but basically it shall be a tuple, right?

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/plugins/ddp_plugin.py:104]: also batch_transfer to a particular device is possible with DDP? isn't this handled by pytorch?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py:111]: It should work fine.

**Comment by tchaton** [comment]
Hey @rohitgr7,

I have added a test. Seems to work fine.

Best,
T.C

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py:104]: It doesn't seem to be handled for custom batches.

**Comment by SeanNaren** [comment]
Thanks @tchaton!

As Adrian said, this will work for single process/single device, but may not work with DDP2 or if device ids are specified. Could you add a test to check if this is the case? DDP2 on one machine boils down to DP.

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 4]
```diff
@@ -91,23 +92,26 @@ def init_ddp_connection(
...
 
         Args:
-            args: Inputs to the model.
+            batch: Inputs to the model.
             model: Model to train.
-        Returns: args moved to correct device if needed.
+        Returns: batch moved to correct device if needed.
         """
-        return args
+        if model.device_ids is not None and len(model.device_ids) == 1:
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/ddp_plugin.py:111]: This is a nice addition! Any reason why we can't make this a property `is_single_process_single_device`? I think it also clears things up and maybe even a comment:

```
# when using single process/single device move all objects to appropriate device as no scatter is necessary
```

File: pytorch_lightning/overrides/data_parallel.py [hunk 5]
```diff
@@ -220,6 +220,11 @@ def reducer_prepare_for_backwards(self, output):
     def reducer_reset_hooks(self):
         self._reducer_prepared_for_backwards = False
 
+    @property
+    def running_single_process_per_device(self) -> bool:
```
Comment (SeanNaren) [inline pytorch_lightning/overrides/data_parallel.py:224]: @awaelchli I think this is fine, in your refactor you've made this a wrapper module. We can move this logic to the wrapper right?  https://github.com/PyTorchLightning/pytorch-lightning/pull/5185/files#diff-d9eba1c616cf4db7a99b8fcdb0e97f9de7b92df25eba32ca0a22f51993735c57R155

File: pytorch_lightning/plugins/sharded_plugin.py [hunk 6]
```diff
@@ -36,15 +36,14 @@ def configure_ddp(
             self, model: LightningModule, device_ids: List[int]
     ):
         self._wrap_optimizers(model)
-        return LightningShardedDataParallel(model, sharded_optimizer=model.trainer.optimizers)
+        model = LightningShardedDataParallel(model, sharded_optimizer=model.trainer.optimizers)
+        model.device_ids = device_ids
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/sharded_plugin.py:40]: I think this is fine, how about `self.device_ids` and keeping track within the plugin rather than the model? Same with ddp_plugin.py

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/plugins/ddp_plugin.py:109]: so for normal batches (tensors..), this still works if we put the batches on the correct device before pytorch does in https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html??

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py]: Done !

File: pytorch_lightning/plugins/sharded_plugin.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/plugins/sharded_plugin.py]: Done !

File: pytorch_lightning/accelerators/ddp2_accelerator.py [hunk 7]
```diff
@@ -72,7 +72,7 @@ def test_step(self, args):
         return self._step(args)
 
     def _step(self, args):
-        args = self.ddp_plugin.on_before_forward(self.trainer.get_model(), *args)
+        args = self.ddp_plugin.on_before_forward(self.trainer.model, *args)
```
Comment (carmocca) [inline pytorch_lightning/accelerators/ddp2_accelerator.py:75]: question: Why this change?

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 8]
```diff
@@ -91,23 +93,30 @@ def init_ddp_connection(
...
 
         Args:
-            args: Inputs to the model.
+            batch: Inputs to the model.
             model: Model to train.
-        Returns: args moved to correct device if needed.
+        Returns: batch moved to correct device if needed.
         """
-        return args
+        if self.is_running_single_process_per_device:
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/ddp_plugin.py:116]: If you could add a lil explanation as to why this is needed here that would be awesome:

```
objects do not need to be scattered in single process per device, move objects upfront to device 
```

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py:118]: Done !

File: pytorch_lightning/accelerators/ddp2_accelerator.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/ddp2_accelerator.py]: Great catch ! Reverted back !

File: tests/models/test_hooks.py [hunk 9]
```diff
@@ -124,6 +125,49 @@ def transfer_batch_to_device(self, data, device):
...
+    def collate_fn(batch):
+        return CustomBatch(batch)
+
+    class TestModel(BoringModel):
+        def training_step(self, batch, batch_idx):
+            assert batch.samples.device == self.device
+            assert isinstance(batch_idx, int)
+
+        def train_dataloader(self):
+            return torch.utils.data.DataLoader(RandomDataset(32, 64), collate_fn=collate_fn)
```
Comment (carmocca) [inline tests/models/test_hooks.py:154]: It's a bit weird the way you are using `collate_fn` here. I think what you want is to have `CustomBatch` subclass `RandomDataset`

File: tests/special_tests.sh [hunk 10]
```diff
@@ -20,3 +20,4 @@ python ${DEFAULTS} tests/plugins/test_ddp_sequential_plugin.py::test_ddp_sequent
 python ${DEFAULTS} tests/plugins/test_ddp_sequential_plugin.py::test_ddp_sequential_plugin_ddp_rpc_manual_amp
 python ${DEFAULTS} tests/plugins/test_ddp_sequential_plugin.py::test_ddp_sequential_plugin_ddp_rpc_automatic
 python ${DEFAULTS} tests/trainer/logging_tests/test_train_loop_logging_1_0.py::test_logging_sync_dist_true_ddp
+python ${DEFAULTS} tests/models/test_hooks.py::test_transfer_batch_hook_ddp
```
Comment (carmocca) [inline tests/special_tests.sh:23]: question: why is it "special"?

**Comment by rohitgr7** [comment]
does it fix https://github.com/PyTorchLightning/pytorch-lightning/issues/2350?

**Comment by tchaton** [comment]
> does it fix #2350?

Only for DDP.

File: tests/special_tests.sh (hunk 10 above)
Comment (tchaton) [reply tests/special_tests.sh:23]: DDP tests are run using sh tests/special_tests.sh as they don't behave well with pytest.

**Review by carmocca — APPROVED** [review]

**Comment by rohitgr7** [comment]
update the docs too?
https://github.com/PyTorchLightning/pytorch-lightning/blob/dd442b6d335a5553961e7904cd9454b738cb72cd/pytorch_lightning/core/hooks.py#L541-L546

**Comment by tchaton** [comment]
> update the docs too?
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/dd442b6d335a5553961e7904cd9454b738cb72cd/pytorch_lightning/core/hooks.py#L541-L546

Great catch ! Done !

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/ddp_plugin.py:111]: @rohitgr7 this DDP from pytorch is restricted to a few primitives (tensors and dict I believe). DDP actually does more than we  use it for here in Lightning, it supports multiple device indices for which a batch would have to be scattered. For non-tensor objects, generally, the batch size (and batch dim) is undefined therefore scattering is not possible (and gathering for that matter). However, in Lightning we never pass multiple device indices to DDP. Instead, we launch multiple processes and in each we pass a single device indes. This usage is also recommended by pytorch. It also means we don't have to scatter any batches. This is how I understand it, hope it is clearer. 

For DP the issue remains, because there it's always required to scatter and gather (multiple devices).

Comment (rohitgr7) [reply pytorch_lightning/plugins/ddp_plugin.py:111]: ah.. got it now.. thanks @awaelchli . Greate explanation

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 11]
```diff
@@ -91,23 +93,31 @@ def init_ddp_connection(
...
         Override to handle custom input to device logic. For DDP, no logic is required as this is handled internally
         within the DDP wrapper.
 
         Example::
 
-            def on_before_forward(self, model, *args):
-                batch, batch_idx = args
+            def on_before_forward(self, model, *batch):
+                batch, batch_idx = batch
                 return batch.to(model.device)
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/ddp_plugin.py:111]:
```suggestion
                batch, batch_idx = batch
                return batch.to(model.device), batch_idx
```
btw it doesn't always contain only two values, can have `opt_idx`, `dataloader_idx`, `hiddens` as well. maybe generalize this example accordingly? Also, why rename it from args to batch?

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 12]
```diff
@@ -91,23 +93,31 @@ def init_ddp_connection(
...
         Args:
-            args: Inputs to the model.
+            batch: Inputs to the model.
             model: Model to train.
-        Returns: args moved to correct device if needed.
+        Returns: batch moved to correct device if needed.
         """
-        return args
+        if self.is_running_single_process_per_device:
+            batch = model.transfer_batch_to_device(batch, model.device)
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/ddp_plugin.py:119]:
```suggestion
            batch = model.transfer_batch_to_device(batch)
```
default is `model.device`

Comment (rohitgr7) [reply pytorch_lightning/plugins/ddp_plugin.py:119]: in the case of `truncated_bptt_steps`, `hiddens` will be moved to the device again?

File: tests/models/test_hooks.py [hunk 13]
```diff
@@ -124,6 +125,49 @@ def transfer_batch_to_device(self, data, device):
...
+
+    model = TestModel()
+    model.validation_step = None
+    model.training_epoch_end = None
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
```
Comment (rohitgr7) [inline tests/models/test_hooks.py:161]: DDP with 1 batch?

File: pytorch_lightning/core/hooks.py [hunk 14]
```diff
@@ -539,9 +540,9 @@ def transfer_batch_to_device(self, batch, device)
             any other device than the one passed in as argument (unless you know what you are doing).
 
         Note:
-            This hook only runs on single GPU training (no data-parallel). If you need multi-GPU support
-            for your custom batch objects, you need to define your custom
-            :class:`~torch.nn.parallel.DistributedDataParallel` or
+            This hook only runs on single GPU training and DDP.
+            If you need multi-GPU support for your custom batch objects in `dp` or `ddp2`,
+            you need to define your custom :class:`~torch.nn.parallel.DistributedDataParallel` or
```
Comment (rohitgr7) [inline pytorch_lightning/core/hooks.py:545]:
```suggestion
            If you need multi-GPU support for your custom batch objects in ``dp`` or ``ddp2``,
            you need to define your custom :class:`~torch.nn.parallel.DistributedDataParallel` or
```

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py:111]: Done

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py:119]: You are right. I need to check this.

File: tests/models/test_hooks.py (hunk 13 above)
Comment (tchaton) [reply tests/models/test_hooks.py:161]: Why do you think it could be a problem ?

Comment (tchaton) [reply tests/models/test_hooks.py:161]: set to 2

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 15]
```diff
@@ -100,13 +108,15 @@ def on_before_forward(self, model: LightningModule, *args):
...
                 batch, batch_idx = args
-                return batch.to(model.device)
+                return batch.to(model.device), batch_idx
 
         Args:
-            args: Inputs to the model.
+            batch: Inputs to the model.
             model: Model to train.
-        Returns: args moved to correct device if needed.
+        Returns: batch moved to correct device if needed.
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/ddp_plugin.py:116]:
```suggestion
        Returns: args moved to correct device if needed.
```

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 16]
```diff
@@ -100,13 +108,15 @@ def on_before_forward(self, model: LightningModule, *args):
 
             def on_before_forward(self, model, *args):
                 batch, batch_idx = args
-                return batch.to(model.device)
+                return batch.to(model.device), batch_idx
 
         Args:
-            args: Inputs to the model.
+            batch: Inputs to the model.
             model: Model to train.
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/ddp_plugin.py:115]:
```suggestion
            model: Model to train.
            args: Inputs to the model.
```

**Review by rohitgr7 — APPROVED** [review]
lgtm

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 17]
```diff
@@ -91,6 +93,12 @@ def init_ddp_connection(
...
 
+    @property
+    def is_running_single_process_per_device(self) -> bool:
+        # objects do not need to be scattered in single process per device, move objects upfront to device
+        # used in `self.on_before_forward`
+        return self.device_ids is not None and len(self.device_ids) == 1
+
     def on_before_forward(self, model: LightningModule, *args):
         """
         Override to handle custom input to device logic. For DDP, no logic is required as this is handled internally
```
Comment (awaelchli) [inline pytorch_lightning/plugins/ddp_plugin.py:104]: this is now outdated

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 18]
```diff
@@ -59,15 +60,16 @@ def configure_ddp(self, model, device_ids):
...
         self._ddp_kwargs["find_unused_parameters"] = self._ddp_kwargs.get(
-            "find_unused_parameters", True
+            "find_unused_parameters", False
         )
         model = LightningDistributedDataParallel(
             model,
             device_ids=device_ids,
             **self._ddp_kwargs,
         )
+        self.device_ids = device_ids
```
Comment (awaelchli) [inline pytorch_lightning/plugins/ddp_plugin.py:72]: not sure about this, setting the device ids in every subclass is not so convenient

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 17 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py:104]: Done !

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 18 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py:72]: Done !

**Comment by carmocca** [comment]
Can you update the CHANGELOG? (after #5378 is merged)

**Review by awaelchli — APPROVED** [review]
not liking the device_ids assignment but we will be able to handle that more elegantly through the refactor. otherwise LGTM thanks for taking a look at this.

**Review by williamFalcon — APPROVED** [review]
